### PR TITLE
Metal: Improve startup times by using concurrent shader compilation APIs

### PR DESCRIPTION
**Pull request by stuartcarnie**
This PR dramatically reduces startup times for Godot when Metal shaders must be recompiled. 

The two primary changes are:

- Only schedule one `MTLLibrary` compilation request per unique shader source
- Use the [concurrent `MTLLibrary` creation API](https://developer.apple.com/documentation/metal/mtldevice/1433351-newlibrarywithsource?language=objc), to allow multiple compilation requests to run in the background

> [!NOTE]
>
> The number of concurrent compilation requests is managed by the OS, and is enabled in Godot by enabling the [`shouldMaximizeConcurrentCompilation`](https://developer.apple.com/documentation/metal/mtldevice/4133938-shouldmaximizeconcurrentcompilat?language=objc) property.

I ran the Bistro demo, which is a great stress test, and cleared the _entire_ Metal shader cache. My hardware is a MacBook Pro M1 Max. For comparison, starting MoltenVK under these conditions takes about 48 seconds for the UI to become responsive.

| Before | After |
| :---- | :---- |
| 2m:30s | 30s |

Relaunching after cache is warmed is about 8-9 seconds.

## Testing

I tested absolute worst-case scenarios, by removing the entire Metal shader cache for Godot. Normally, various restarts of Godot would compile core shaders, such as for the editor, so that the impact of launching the Bistro demo in the editor is not as bad.

You can find the OS Metal shader cache using the following command

```sh
ls `getconf DARWIN_USER_CACHE_DIR`/org.godotengine.godot
```

which might look similar to the following:

```
total 0
drwxr-xr-x@ 8 stuartcarnie  staff  256 27 Dec  2023 com.apple.gpuarchiver
drwxr-xr-x@ 7 stuartcarnie  staff  224 18 Jul 06:21 com.apple.metal
drwxr-xr-x@ 3 stuartcarnie  staff   96 10 Mar  2023 com.apple.metalfe
```

And to clear the cache, remove all those directories:

```sh
rm -rf `getconf DARWIN_USER_CACHE_DIR`/*
```

### Before

Launching the current version of Godot with Metal, and after adding some instrumentation that can be visualised in Instruments, we can see all the compilation is serial:

CleanShot 2024-08-25 at 08 14 14@2x

It takes about 2m:30s for the UI to become responsive.

### After

Clearing the cache and running with the improvements, UI is responsive at around 30s. The following chart shows that shader compilation is now concurrent.

CleanShot 2024-08-25 at 08 26 16@2x

Take note of the section summary, for the **shader_compile** statistics, that indicate **936** unique shaders were compiled.

It turns out that only 250 shaders (26%) of the requested shaders are used by the editor. I implemented an alternative caching strategy that only compiles the shader when creating a pipeline to validate this:

CleanShot 2024-08-25 at 08 30 02@2x

This strategy can be tested by specifying the following environment variable when launching Godot:

```
GODOT_MTL_SHADER_LOAD_STRATEGY=lazy
```
 
> [!IMPORTANT]
>
> I expect this strategy will be better if Godot introduces concurrent pipeline creation / compilation

### CODE REVIEW

**Comment by Calinou** [comment]
> I expect this strategy will be better if Godot introduces concurrent pipeline creation / compilation

Is this what https://github.com/godotengine/godot/pull/90400 introduces?

**Comment by stuartcarnie** [comment]
> > I expect this strategy will be better if Godot introduces concurrent pipeline creation / compilation
> 
> Is this what #90400 introduces?

If #90400 creates render and compute pipelines on separate threads, concurrently, then yes 👍🏻  I will be able to easily verify it, once it lands.

**Comment by passivestar** [comment]
Did a quick test with bistro without the cache and got similar results:

| Master Metal | Master Vulkan | This PR Metal |
|--------|--------|--------|
| 2m 30s | 1m 24s | 35s  🎉 |

This PR with cache: 8 seconds

**Comment by stuartcarnie** [comment]
Great news!

**Comment by DarioSamo** [comment]
> If #90400 creates render and compute pipelines on separate threads, concurrently, then yes 👍🏻 I will be able to easily verify it, once it lands.

Yeah it does. You might want to make sure it doesn't end up in a situation where it breaks the expectation that it handles the MT at the higher level and therefore ends up doing double the duty of thread management once it does.

File: drivers/metal/metal_objects.h
Comment (DarioSamo) [reply drivers/metal/metal_objects.h:505]: Probably worth mentioning this exact behavior is introduced by #90400 but at a general level for all APIs.

**Comment by stuartcarnie** [comment]
> Yeah it does. It can create both pipelines and shaders by maximizing the threads available. You might want to make sure it doesn't end up in a situation where it breaks the expectation that it handles the MT at the higher level and therefore ends up doing double the duty of thread management once it gets merged.

@DarioSamo Awesome!

It doesn't, as it calls the synchronous API when in `LAZY` mode, which we'll use one #90400 merges. I'll just change the default behaviour so that it only compiles the `MTLLibrary` when creating the pipeline. I expect the OS will still queue up compilation requests, so we'll be able to issue them in parallel on different threads.

This will be a huge win overall, as we'll only be compiling the libraries for the pipelines we're actually using.

File: drivers/metal/metal_objects.h
Comment (stuartcarnie) [reply drivers/metal/metal_objects.h:505]: Thats great!

**Review by akien-mga — APPROVED** [review]
Only did a cursory check, but it's well scoped to the Metal driver so I think I can defer to Stuart and merge (Clay is away for a week). This will be good to have in the imminent 4.4.dev1 snapshot to give testers a better first impression of the new Metal driver.

**Comment by akien-mga** [comment]
Thanks! Great work :tada:

File: drivers/metal/rendering_device_driver_metal.mm
Comment (zeux) [inline drivers/metal/rendering_device_driver_metal.mm:75]: Was this meant to be hardcoded?

Comment (stuartcarnie) [reply drivers/metal/rendering_device_driver_metal.mm:75]: No, I'm switching that to `org.godotengine.godot.metal` in a follow-up PR

Comment (akien-mga) [reply drivers/metal/rendering_device_driver_metal.mm:75]: Heh good catch, I was eager to merge this for 4.4.dev1 so I overlooked it.
Stuart gets one dev snapshot of fame in the log uid ;)

Comment (stuartcarnie) [reply drivers/metal/rendering_device_driver_metal.mm:75]: I've correct that in #96089 too 😊

Comment (stuartcarnie) [reply drivers/metal/rendering_device_driver_metal.mm:75]: Thanks for the spot, @zeux!

**Pull request merged**
